### ISSUE TITLE
proposal: x/mod: new repo for module mechanics

### ISSUE THREAD

**Issue by rsc** [golang/go#31761]
For the initial development of module support, we've kept everything in internal directories to make it easier to make changes as our understanding of what the pieces should look like becomes clearer. But now I think the basic pieces have become pretty clear, and we should think about exporting some packages to help other people who want to write tools working directly with module mechanics. 

Note that I said "module mechanics". This is about supporting direct manipulation of modules themselves. It is _not_ about supporting general development tools that need to do things like load packages in module mode. That use case, where modules are incidental rather than the focus, should remain in x/tools, specifically x/tools/go/packages. And the specific case of loading packages should still be done by invoking the go command, which will remain the single point of truth for those algorithms.

I propose to create a new repo golang.org/x/mod that will initially contain:

 - **golang.org/x/mod/module**. defines type module.Version (struct {Path, Version string}) and !-encoding of paths and versions (from [cmd/go/internal/module](https://godoc.org/cmd/go/internal/module))
 - **golang.org/x/mod/modfile**. go.mod file reader, editor (from [cmd/go/internal/modfile](https://godoc.org/cmd/go/internal/modfile))
 - **golang.org/x/mod/semver**. semantic version parser (from [cmd/go/internal/semver](https://godoc.org/cmd/go/internal/semver))
 - **golang.org/x/mod/sumdb**. go checksum database client and server protocol (from [golang.org/x/exp/sumdb/internal/sumweb](https://godoc.org/golang.org/x/exp/sumdb/internal/sumweb))
 - **golang.org/x/mod/sumdb/storage**. go checksum database storage layer (from [golang.org/x/exp/sumdb/internal/tkv](https://go-review.googlesource.com/c/exp/+/161662), [database](https://go-review.googlesource.com/c/exp/+/161665))
 - **golang.org/x/mod/sumdb/tlog**. transparent log (from [golang.org/x/exp/sumdb/internal/tlog](https://godoc.org/golang.org/x/exp/sumdb/internal/tlog))
 - **golang.org/x/mod/sumdb/note**. signed note formats (from [golang.org/x/exp/sumdb/internal/note](https://godoc.org/golang.org/x/exp/sumdb/internal/note))
 - **golang.org/x/mod/sumdb/dirhash**. definition of `h1:` directory tree hash used in go.sum file (from [cmd/go/internal/dirhash](https://godoc.org/cmd/go/internal/dirhash))
 - **golang.org/x/mod/sumdb/googlecloud...**. sub-module with basic google cloud server bindings (from [golang.org/x/exp/sumdb/internal/googlecloud...](https://go-review.googlesource.com/c/exp/+/161664))
 - **golang.org/x/mod/proxy**. go module proxy client and server protocol (like [x/mod/sumdb](https://godoc.org/golang.org/x/exp/sumdb/internal/sumweb); not written yet)
 - **golang.org/x/mod/index**. go module index client and server protocol (like [x/mod/sumdb](https://godoc.org/golang.org/x/exp/sumdb/internal/sumweb); not written yet)
 - **golang.org/x/mod/gosumcheck**. command: go.sum checker (from [golang.org/x/exp/sumdb/gosumcheck](https://godoc.org/golang.org/x/exp/sumdb/gosumcheck))
 - **golang.org/x/mod/gomodproxy**. command: trivial go module proxy for reference and trivial uses (not published yet; maybe also “go mod proxy”)

There's an open issue that this proposal addresses (about exposing some of this stuff) but once again I cannot find it.

/cc @bcmills @jayconrod @ianthehat @myitcv @rogpeppe

**Comment by bcmills** [comment]
> There's an open issue that this proposal addresses (about exposing some of this stuff) but once again I cannot find it.

That's #28101.

**Comment by dmitshur** [comment]
> I propose to create a new repo golang.org/x/mod

Does this proposal include a plan for how the new `module`, `modfile`, `semver`, etc., packages will be kept in sync with the ones in `cmd/go/internal/...`?

**Comment by bcmills** [comment]
I would assume that they'll be vendored in using `go mod vendor`.

That does imply that we'll have to have a clean (and perhaps smaller) API boundary between `golang.org/x/mod` and `cmd/go/internal`, but that's arguably a good thing anyway.

**Comment by rsc** [comment]
@dmitshur, yes they would be vendored into cmd/vendor using "go mod vendor".
All the ones I've chosen should already have clean API boundaries, and if not, we'd work on that.

**Comment by ianthehat** [comment]
I need quite a few of these in order to support mod files in editors. I have an experiment I have been working on that has them copied to an internal directory of x/tools, but it would be much nicer if I could go straight to using this repository instead.

**Comment by hyangah** [comment]
Consider pseudo-version related logic as well
 e.g. https://go.googlesource.com/go/+/refs/heads/master/src/cmd/go/internal/modfetch/pseudo.go

**Comment by jayconrod** [comment]
The list of packages above looks good. I expect this will be useful for developing `gorelease` and a lot of other tools.

I notice that none of `internal/modload`, `internal/modfetch`, `internal/modfetch/coderepo`, `internal/mvs` are in the list. I think that's the right decision for now -- these are all pretty tightly coupled with cmd/go. It might be worth moving these to x/mod after some refactoring though. I think the proxy and discovery projects have run into use cases where query and download functionality would be useful, and the Go command doesn't do quite what they need.

**Comment by rsc** [comment]
Sentiment here seems overwhelmingly in favor. Accepting proposal.

**Comment by rsc** [comment]
Obviously x/mod itself is "unreleased" but it will have effects in the main repo, so milestoned to Go 1.13.

**Comment by rsc** [comment]
@andybons, I created this repo. Can you take care of adding it to gopherbot, builders, etc.? Thanks.

**Comment by andybons** [comment]
Everything is set up for x/mod (mirroring, gerritbot, golang.org/x/mod, etc.)

**Comment by dmitris** [comment]
I'm looking for a published library with the types / structures listed in the output of `go mod help edit` - namely `Module`, `GoMod`, `Require`, and `Replace`, ideally with the `String()` methods attached?  Would they be in the new library (`x/mod`)?

**Comment by dmitshur** [comment]
@dmitris Yes, that will be the `golang.org/x/mod/modfile` package. It hasn’t been added to x/mod just yet.

**Comment by tsal** [comment]
Do we know when that might happen? This would be very useful for some CI tooling.

**Comment by tsal** [comment]
Oops! I missed that this was added to the Go1.13 milestone back in May!

**Comment by rsc** [comment]
This is done.
